File: src/main/java/mip/examples/Knapsack.java
package mip.examples;

import mip.*;

import java.util.*;

public class Knapsack {

    public static void main(String[] args) {
        int[] p = { 10, 13, 18, 31, 7, 15 };
        int[] w = { 11, 15, 20, 35, 10, 33 };
        int c = 47;
        int n = w.length;

        // creating model
        Model m = new Model("Knapsack");

        // creating vars and setting objective function
        List<Var> x = new ArrayList<>();
        for (int i = 0; i < n; i++)
            x.add(m.addBinVar("item_" + i));

        // creating and setting objective function
        LinExpr obj = new LinExpr();
        for (int i = 0; i < n; i++)
            obj.add(x.get(i), p[i]);
        m.setObjective(MIP.maximize(obj));

        // adding constraint
        LinExpr expr = new LinExpr();
        for (int i = 0; i < n; i++)
            expr.add(x.get(i), w[i]);
        m.addLeq(expr, c);

        // solving model
        m.optimize();

        // printing result
        System.out.print("Selected items: { ");
        for (int i = 0; i < n; i++)
            if (x.get(i).getX() >= 0.99)
                System.out.print(i + " ");
        System.out.println("}");
    }
}

File: src/main/kotlin/mip/examples/Knapsack.kt
package mip.examples

import mip.*
import kotlin.math.roundToInt

private

fun main() {
    // problem's input data
    val p = arrayOf(10, 13, 18, 31, 7, 15)
    val w = arrayOf(11, 15, 20, 35, 10, 33)
    val c = 47
    val I = w.indices // 0 until w.size

    // creating model
    val m = Model("Knapsack", solverName = GUROBI)

    // creating vars and setting objective function
    val x = I.map { i -> m.addBinVar("item_$i") }
    m.objective = maximize(I.sum { i -> p[i] * x[i] })

    // adding constraint and solving model
    m += I.sum { i -> w[i] * x[i] } leq c
    m.optimize()
    m.write("queens.lp")

    // printing result
    val selected = x.filter { it.x >= 0.99 }
    println("Total cost: ${m.objectiveValue}")
    println("Selected items: $selected")

    // sanity tests
    assert(m.status == OptimizationStatus.Optimal)
    assert(m.objectiveValue.roundToInt() == 41)
    assert(m.constrs[0].slack.roundToInt() == 1)
    assert(m.validateOptimizationResult())
}

fun runKnapsack() = main()